flytekit/image_spec/image_spec.py

```diff
@@ -25,7 +25,7 @@ class ImageSpec:
         name: name of the image.
         python_version: python version of the image. Use default python in the base image if None.
         builder: Type of plugin to build the image. Use envd by default.
-        source_root: source root of the image.
+        source_root: source root of the image, or the dockerfile context path.
```

Comment:
You wrote:

> I'm not a huge fan of this solution, I think it would be nice to not have the plugins installed on the node
The current way is a bit too cluttered for me as it somewhat defeats the purpose of having the plugin abstraction.

I agree with it being to cluttered and defeating the purpose of having a plugin abstraction. Maybe the `register` method of the `ImageBuildEngine` should receive an accompanying `ImageSpec` class for each `ImageSpecBuilder`?
Could you please point me to where the `ImageSpec` is needed at runtime? Maybe we can avoid this by running the relevant code only when the flyte context tells us we are in registration mode?

I think @pingsutw should have the ultimate say in this.

Reply:
For now, source root is always set by flytekit, and it's the first ancestor from source_path that does not contain a `__init__.py` file. Do you want to set it by yourself?